File: ruby/two_sum.rb
# frozen_string_literal: true

# https://leetcode.com/problems/two-sum/
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  sorted_nums = nums.map.with_index { |v, i| [v, i] }.sort_by(&:first)
  two_sum_on_sorted_array(sorted_nums, target)
end

def two_sum_on_sorted_array(array, target)
  i = 0
  j = nums.length - 1

  while i < j
    case array[i].first + array[j].first <=> target
    when 0 then return [array[i].last, array[j].last]
    when 1 then j -= 1
    when -1 then i += 1
    end
  end
end

File: typescript/two-sum.ts
// https://leetcode.com/problems/two-sum/

function twoSum(nums: number[], target: number): [number, number] {
  const met = new Map<number, number>();
  for (let i = 0; i < nums.length; i++) {
    const element = nums[i];
    const complement = target - element;
    if (met.has(complement)) {
      return [i, met.get(complement)!];
    } else {
      met.set(element, i);
    }
  }
  return [-1, -1];
}

import { assertArrayIncludes } from "std/assert/mod.ts";

Deno.test("examples", () => {
  assertArrayIncludes(twoSum([2, 7, 11, 15], 9), [0, 1]);
  assertArrayIncludes(twoSum([3, 2, 4], 6), [1, 2]);
  assertArrayIncludes(twoSum([3, 3], 6), [0, 1]);
});
